File: Data_Structures/javascript/Stack/reverseStackRecursion.js
const reverseStackRecursion = stack => {
    if (!stack.isEmpty()) {
        let tmp = stack.pop();
        reverseStackRecursion(stack);
        insertAtTheBottom(tmp, stack);
    }
}

// Start inserting all the elements from the bottom of our stack
insertAtTheBottom = (tmp, stack) => {
    if (stack.isEmpty()) {
        stack.push(tmp);
    } else {
        let x = stack.pop();
        insertAtTheBottom(tmp, stack);
        stack.push(x);
    }
}

let stack = new Stack();
 stack.push(1);
 stack.push(2);
 stack.push(3);
 stack.push(4);
 stack.push(5);

 reverseStack(stack); //call the function

 //Print the stack
 while(!stack.isEmpty()){
   console.log(stack.pop());
 }

File: Data_Structures/C++/Stack/reverseStackRecursion.cpp
#include <iostream>
#include <stack>

using namespace std;

stack<char> Stack;

string ns;

void    insertBottom(char x)
{
    if (Stack.size() == 0)
        Stack.push(x);
    else {
        char t = Stack.top();
        Stack.pop();
        insertBottom(x);
        Stack.push(t);
    }
}

void    reverseStackRecursion()
{
    if (Stack.size() > 0) {
        char tmp = Stack.top();
        Stack.pop();
        reverseStackRecursion();
        insertBottom(tmp);
    }
}

int main()
{
    Stack.push('1');
    Stack.push('2');
    Stack.push('3');
    Stack.push('4');

    cout << "Original Stack" << endl;
    cout << "1" << " " << "2" << " " << "3" << " " << "4" << endl;

    reverseStackRecursion();
    cout << "Reversed Stack" << endl;
    while (!Stack.empty()) {
        char p = Stack.top();
        Stack.pop();
        ns += p;
    }
    cout << ns[3] << " " << ns[2] << " " << ns[1] << " " << ns[0] << endl;
    return 0;
}
